lib/schemes/HttpUriPlugin.js

```diff
@@ -0,0 +1,89 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+const { URL } = require("url");
+const NormalModule = require("../NormalModule");
+const memorize = require("../util/memorize");
+
+/** @typedef {import("../Compiler")} Compiler */
+
+class HttpUriPlugin {
+	/**
+	 * Apply the plugin
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		compiler.hooks.compilation.tap(
+			"HttpUriPlugin",
+			(compilation, { normalModuleFactory }) => {
+				const resolveHandler = resourceData => {
+					const url = new URL(resourceData.resource);
+					resourceData.path = url.origin + url.pathname;
+					resourceData.query = url.search;
+					resourceData.fragment = url.hash;
+					return /** @type {true} */ (true);
+				};
+				const readHandler = (scheme, getBuiltin) => (
+					resource,
+					module,
+					callback
+				) => {
+					return getBuiltin().get(new URL(resource), res => {
+						if (res.statusCode !== 200) {
+							res.destroy();
+							return callback(
+								new Error(`${scheme} request status code = ${res.statusCode}`)
+							);
+						}
+
+						const bufferArr = [];
+
+						res.on("data", chunk => {
+							bufferArr.push(chunk);
+						});
+
+						res.on("end", () => {
+							if (!res.complete) {
+								return callback(new Error(`${scheme} request was terminated`));
+							}
+
+							callback(null, Buffer.concat(bufferArr));
+						});
+					});
+				};
+
+				normalModuleFactory.hooks.resolveForScheme
+					.for("http")
+					.tap("HttpUriPlugin", resolveHandler);
+				normalModuleFactory.hooks.resolveForScheme
+					.for("https")
+					.tap("HttpUriPlugin", resolveHandler);
+				NormalModule.getCompilationHooks(compilation)
+					.readResourceForScheme.for("http")
+					.tapAsync(
+						"HttpUriPlugin",
+						readHandler(
+							"http",
+							memorize(() => require("http"))
+						)
+					);
+				NormalModule.getCompilationHooks(compilation)
+					.readResourceForScheme.for("https")
+					.tapAsync(
+						"HttpUriPlugin",
+						readHandler(
+							"https",
+							memorize(() => require("https"))
+						)
+					);
+			}
+		);
+	}
+}
+
+module.exports = HttpUriPlugin;
```

Comment:
Maybe UrlPlugin/UriPlugin? Because we supports and http and https protocols

Reply:
probably we need Http and Https since we could enhance https in future (e.g. provide options to get method: supported TLS version, ciphers etc)

Reply:
It's experimental so we can change it later again. I won't consider this done yet, but I didn't want to remove the code from the PR

Reply:
> It's experimental so we can change it later again
We can change plugins design later, but I still think that naming and amount of plugins better to fix on the coast =)

Also pros for Http and Https is that Node.js have 2 separate modules too

Reply:
ok I splitted it into two plugins